feat(blog): implement contentful source

JulianAnthes opened the pull request:
This PR implements the [gatsby-source-contentful](https://www.gatsbyjs.com/plugins/gatsby-source-contentful/) plugin.

## What's changing?

In the first step only blog posts are moved from the repository (markdown files) to contentful.

## ToDos

- [x] implement plugin
- [x] connect contentful ↔️ gatsby
- [x] create content models on contentful
  - [x] Blog Post – _blog post itself_
  - [x] Blog Post Author – _blog post authors (referenced by blog posts)_
  - [x] Blog Post Hero – _blog post hero images (referenced by blog posts)_
  - [x] Code Block – _code block with language field (referenced by blog posts)_
  - [x] Blog Post Collapsible – _collapsible box with rich text (referenced by blog posts)_
- [x] generate blog post teaser via graphql data
- [x] generate blog posts via graphql data
  - [x] check for missing types/elements
    - [x] links inside rich text
    - [x] marks (`bold`, `italic`, `underline`) 
  - [x] check embedded images (inside rich text)
  - [x] check embedded videos
- [x] rewrite `gatsby-plugin-feed` config in `gatsby-config.ts`
- [x] move existing blog posts to contentful (_see table below_)
- [x] clean up
  - [x] remove `blog-posts` folder
  - [x] update/remove `Writing a blog post` section in `README.md`
  - [x] remove `blog-posts` references from `gatsby-config.ts`
- [x] improvement needed
  - [x] reading time calculation for `ContentfulBlogPost` in `gatsby-config.ts:121`



## Blog Posts Transfer Status

🟢 **Approved**  🟡 **Waiting To Be Approved** 🔴 **Work In Progress**

| File Name                                                | Status | Assigned To | Reviewed By |
|----------------------------------------------------------|--------|-------------|-------------|
| `angular-inject-the-injector.md`                         | 🟢     | @dschu-lab  | @milljoniaer, @georgiee         |
| `angular-workshop-2019.md`                               | 🟢    | @dschu-lab  | @milljoniaer         |
| `boosting-developer-productivity-with-github-actions.md` |  🟢     | @georgiee   | @milljoniaer         |
| `cloudfront-cache-efficiency.md`                         | 🟢     | @georgiee   | @milljoniaer         |
| `daniel-interview.md`                                    | 🟢    | @dschu-lab  | @milljoniaer         |
| `enterprises-benefit-from-scoped-npm-registries.md`      | 🟢      | @dschu-lab  | @milljoniaer, @georgiee         |
| `felix-interview.md`                                     | 🟢  | @georgiee   | @milljoniaer        |
| `four-ways-to-improve-collaboration-in-your-team.md`     | 🟢     | @dschu-lab  | @milljoniaer         |
| `gather-metaverse-bump-each-other-2021.md`               | 🟢    | @dschu-lab  | @milljoniaer, @georgiee         |
| `gatsby-ssr-caching.md`                                  | 🟢     | @dschu-lab  | @milljoniaer         |
| `getting-started-gatsby-next-remix.md`                   | 🟢     | @dschu-lab  | @milljoniaer        |
| `github-billing-dashboard.md`                            | 🟢     | @dschu-lab  | @milljoniaer         |
| `monorepo-code-owner.md`                                 | 🟢     | @dschu-lab  | @milljoniaer         |
| `optimze-gatsby-urls.md`                                 | 🟢     | @dschu-lab  | @milljoniaer, @georgiee        |
| `return-to-office-interview.md`                          |  🟢  | @georgiee   | @milljoniaer, @georgiee        |
| `the-modern-code-reviewer-2021.md`                       | 🟢     | @dschu-lab  | @milljoniaer, @georgiee         |
| `typescript-ast-type-checker.md`                         | 🟢     | @dschu-lab  | @milljoniaer, @georgiee        |

JulianAnthes committed:
docs: add info for apple silicon & env vars

JulianAnthes committed:
build(deps): add gatsby contentful source plugin

JulianAnthes committed:
chore: exclude `.vscode/` from version control

JulianAnthes committed:
feat: configure & activate contentful source plugin

ghost commented:
# [Gatsby Cloud](https://www.gatsbyjs.com) Build Report
**satellytes.com-main**

:tada: Your build was successful! [See the Deploy preview here.](https://satellytescommain21751-featimplementcontentfulsource.gtsb.io)

## Build Details

[View the build logs here.](https://gatsbyjs.com/dashboard/f215130d-15e3-4176-9252-e13166fe48de/sites/674be368-adf0-4ac5-8fae-f9a4493adc51/builds/c0c032ae-00bf-4da0-8e8a-2d7366f09023/details)

:clock1: Build time: **54s**

## Performance

### Lighthouse report

| Metric | Score |
| ----- | ---- |
| Performance |   :large_orange_diamond: 78   |
| Accessibility |   :green_heart: 97   |
| Best Practices |   :green_heart: 100   |
| SEO |   :large_orange_diamond: 79   |

:link: [View full report](https://build-c0c032ae-00bf-4da0-8e8a-2d7366f09023.gtsb.io/reports/lighthouse/index.html)

georgiee commented on README.md:
```diff
@@ -9,6 +9,27 @@ To run the project all you need is:
 - Yarn
 - Node >= 16
 
+### Additional setup on Apple Silicon
```
🙏

georgiee commented on gatsby-config.ts:
```diff
@@ -313,6 +329,10 @@ const gatsbyConfig: GatsbyConfig = {
     },
     'gatsby-plugin-svgr',
+    {
+      resolve: 'gatsby-source-contentful',
```
welcome to the party, contentful ✨

georgiee commented:
Can't wait to see this happen 🤩 By the way I forgot to mention that we still have access to our legacy website that relied on contentful, this might help or not: https://github.com/satellytes/satellytes.com-legacy-2020

JulianAnthes committed:
feat(blog): generate blog teaser from contenful source

JulianAnthes committed:
fix: only create slugs for physical files

WIthout checking for `fileAbsolutePath`, all markdown nodes would be affected. But not every markdown node will end up as its own page and not every MarkdownRemark node is created via file, which can cause createFilePath to fail/error.

JulianAnthes committed:
refactor(blog): remove filesystem imports from /blog-posts

JulianAnthes committed:
feat(blog): add types for contentful models

JulianAnthes committed:
feat(blog): add `leadBoxText` field to `ContentfulBlogPost` type

JulianAnthes committed:
feat(blog): calculate reading time for `ContentfulBlogPost` types

JulianAnthes committed:
refactor(blog): generate blog post pages from contentful source

JulianAnthes committed:
fix(blog): set missing link attributes

JulianAnthes committed:
build(deps): add contentful rich text to html/plain renderer

JulianAnthes committed:
feat(blog): add path field to blog posts

JulianAnthes committed:
feat(blog): add plainText & rssHtml resolvers to blog post nodes

JulianAnthes committed:
refactor(blog): improve reading time calculation

JulianAnthes committed:
feat(blog): create excerpt node for blog posts

JulianAnthes committed:
refactor(blog): generate feed from contentful blog posts

JulianAnthes committed:
refactor(blog): use seoMetaText for blog rss feed description

JulianAnthes committed:
refactor(blog): remove excerpt resolver

feedm3 commented on gatsby-node.ts:
```diff
@@ -23,14 +24,28 @@ export const onCreateNode: GatsbyNode<SatellytesNode>['onCreateNode'] = async (
-  if (node.internal.type === `MarkdownRemark`) {
+  if (
+    node.internal.type === `MarkdownRemark` &&
+    Object.keys(node).includes('fileAbsolutePath')
```
Is it because the Markdown types in a Contentful object don't have an absolute path? A comment above would be helpful here.

JulianAnthes replied:
@feedm3 Yes, your assumption is correct. Added a small note.

JulianAnthes committed:
refactor(markdown): split figcaption props

feedm3 commented on gatsby-node.ts:
```diff
@@ -23,14 +24,28 @@ export const onCreateNode: GatsbyNode<SatellytesNode>['onCreateNode'] = async (
+    createNodeField({
+      name: 'path',
+      node,
+      value: `/blog/${node.slug}/`,
```
I really like that the path gets pinned to the `ContentfulBlogPost`. It feels like the `/blog` can be a constant, but I have no idea where to put this constant so it's fine to just leave it as it is 😄

feedm3 commented on gatsby-node.ts:
```diff
@@ -94,3 +109,31 @@ export const onCreateWebpackConfig: GatsbyNode['onCreateWebpackConfig'] =
+
+export const createResolvers = ({ createResolvers }) => {
+  createResolvers({
+    ContentfulBlogPost: {
```
Idea: Would it make sense to also add the `path` here? To not have 2 places where we modify the type.

JulianAnthes replied:
Since the `slug` is generated during `onCreateNode` I tended to include the `path` there as well.

JulianAnthes committed:
feat(blog): include ContentfulAssets in GraphQL query

JulianAnthes committed:
feat(blog): render embedded assets

feedm3 commented on gatsby/create-pages/create-blog-posts.ts:
```diff
@@ -13,44 +12,41 @@ export const createBlogPosts = async ({
-        };
+        id: string;
+        publicationDate: string;
+        slug: string;
```
We don't need the `slug` here, you can use your `path` variable.

feedm3 commented on src/components/pages/blog-post/blog-post.tsx:
```diff
@@ -1,19 +1,25 @@
+import { LeadboxProps } from '../../content/leadbox/leadbox';
 import { LONG_DATE_FORMAT, useLocaleFormat } from '../../i18n-helpers';
+import { Layout } from '../../layout/layout';
+import cSC from '../../legacy/markdown/custom-components';
```
Very hard to read, maybe use a longer name?

JulianAnthes replied:
Renaming to `customComponents` 🎉

feedm3 commented on src/components/pages/landingpage/blog.tsx:
```diff
@@ -33,18 +33,18 @@ export const Blog = ({ posts }: BlogProps) => {
-              title={item.frontmatter.title}
-              linkTo={item.frontmatter.path}
+              title={item.title}
+              linkTo={`blog/${item.slug}`}
```
`item.path` should also work here, right?

JulianAnthes replied:
Good catch!

feedm3 commented on src/pages/index.tsx:
```diff
@@ -69,28 +69,25 @@ export const IndexPageQuery = graphql`
-      sort: { fields: frontmatter___date, order: DESC }
-      filter: { fileAbsolutePath: { regex: "/(blog-posts)/" } }
+    allContentfulBlogPost(
+      sort: { fields: publicationDate, order: DESC }
```
Very cool 👍

feedm3 reviewed (approved):
I really like the implementation. The architecture of the blog post content is way better than expected. Good work 👍 

I had some small improvements, nothing spectacular. Will already approve.

JulianAnthes committed:
refactor(blog): use path field for page creation

JulianAnthes committed:
refactor(blog): use path field for blog teaser links

JulianAnthes committed:
refactor(blog): rename import into a more meaningful name

JulianAnthes committed:
refactor(blog): use path field for blog teaser

JulianAnthes committed:
fix(markdown): type missmatch when spreading arguments

JulianAnthes committed:
docs: describe why nodes without absolute file path are excluded

JulianAnthes committed:
feat(blog): implement video asset support

JulianAnthes committed:
feat(blog): implement blog post collapsible rendering

georgiee commented on gatsby/create-pages/create-blog-posts.ts:
```diff
@@ -13,44 +12,45 @@ export const createBlogPosts = async ({
-        sort: { order: DESC, fields: [frontmatter___date] }
-        filter: { fileAbsolutePath: { regex: "/(blog-posts)/" } }
-      ) {
+      allContentfulBlogPost(sort: { fields: publicationDate, order: DESC }) {
```
welcome contentful. this looks great 👍

georgiee commented on src/components/content/heroes/blog-hero.tsx:
```diff
@@ -16,7 +17,7 @@
 type ImageHeroProps = Partial<HeroWithText> & {
   image: IGatsbyImageData;
-  attribution?: any;
+  attribution?: Omit<ContentfulBlogPostHero, 'image'>;
```
this is 100% correct but I don't like the semantic it introduces when reading the code. It emphasizes on `image` while it's exactly the opposite. For the reader it's not clear and I think a dervice type `BlogAttribution` would be much better to read the code. Something like 

`type BlogAttribution = Omit<ContentfulBlogPostHero, 'image'>;`

or even better explicitly pick the interesting fields because I thin kwhenever something is added it's mostly NOT attribution related so the assumption of omiting is wrong here, isn't it?

georgiee commented on src/components/pages/blog-post/blog-post.tsx:
```diff
@@ -27,45 +33,148 @@ const PanelContainer = styled.div`
 `;
 
-export const BlogPostPage = ({ markdown, breadcrumb }: BlogPostPageProps) => {
+const contentfulRenderOptions: Options = {
```
Note to ourselves, this makes a great separate file to collect all the wonderful mappings here. I'm still amazed how this is very similar to our rehype approach with markdown. Like we picked the perfect solution before even knowing the contentful migration will benefit from it so much. Superb job 🙌

JulianAnthes replied:
I was also surprised how smooth that one went. Was basically plug&play. 😀

georgiee commented on src/components/pages/blog-post/blog-post.tsx:
```diff
@@ -27,45 +33,148 @@ const PanelContainer = styled.div`
+        /**
+         * Blog Post Collapsible
+         */
+        case 'ContentfulBlogPostCollapsible': {
```
nice my beloved summary/detail toggles are covered too ✨

georgiee commented on src/components/pages/blog-post/blog-post.tsx:
```diff
@@ -1,19 +1,25 @@
-import { MarkdownAst } from '../../legacy/markdown/markdown-ast';
-import SharePanel from './share-panel';
-import FollowPanel from './follow-panel';
+import { renderRichText } from 'gatsby-source-contentful/rich-text';
```
Great that they ship the tools we need

georgiee commented on src/pages/blog.tsx:
```diff
@@ -45,31 +39,27 @@ export default Blog;
+        teaserText
+        slug
+        publicationDate
+        heroImage {
```
thanks for considering a better name in contentful and here 🙏

georgiee commented on src/templates/blog-post.tsx:
```diff
@@ -100,4 +165,24 @@ export const BlogPostPageQuery = graphql`
+  };
+  heroImage: {
+    fullImage: IGatsbyImageData;
+    squaredImage: IGatsbyImageData;
```
This is obsolete now, isn't it after the recent change of the blog hero with the new flag `naturalHeight` ?

georgiee commented on gatsby-config.ts:
```diff
@@ -115,6 +120,8 @@ const gatsbyConfig: GatsbyConfig = {
       options: {
         types: {
           MarkdownRemark: (source) => source.rawMarkdownBody,
+          ContentfulBlogPost: (blogPost) =>
+            documentToPlainTextString(JSON.parse(blogPost.content.raw)),
```
that looks so easy, cool that you find such a simple way

georgiee commented on gatsby-config.ts:
```diff
@@ -312,6 +309,10 @@ const gatsbyConfig: GatsbyConfig = {
       },
     },
     'gatsby-plugin-svgr',
+    {
```
🤘

georgiee reviewed (approved):
Absolutely awesome change. I have not found many things, it just works 💫 

Let's focus on the migration of the blog posts now before merging and once that is done (which is kind of another test for the implementation) we can merge that beauty 🤩

I would also suggest that we check the seo parts with metatags.io just to make sure that those hidden information are still assembled correctly. 

Thank you!

JulianAnthes committed:
fix(blog): add missing div wrapper

this fixes some visual errors, like hr tags not expanding

JulianAnthes committed:
refactor(blog): add BlogAttribution type definition

JulianAnthes committed:
refactor(blog): use attribution type & explicitly check for properties

georgiee commented:
Notes during content migration:
+ We might want to add a few fields to code blocks for additional flags, I just found `yaml{numberLines: true}` as a code marker which enables line numbers.

JulianAnthes committed:
refactor(blog): remove squared image query from hero image

JulianAnthes committed:
refactor(blog): only render author summary when available

JulianAnthes committed:
feat(blog): link images to their sources

JulianAnthes committed:
feat(blog): wrap text on line breaks

This enables line breaks set via Shift+Enter inside rich text fields (Contentful)

JulianAnthes committed:
docs(blog): update blog post creation information

JulianAnthes committed:
chore(blog): remove unused blog post data

georgiee commented:
1. Review Result
https://satellytes.com/blog/enterprises-benefit-from-scoped-npm-registries/

<details>
<summary>GIF  Preview</summary>


![Mar-29-2022 07-02-21](https://user-images.githubusercontent.com/1701755/160536948-b057d2c6-3bd0-45fa-be7a-5c851f92fb4c.gif)
![Mar-29-2022 07-02-34](https://user-images.githubusercontent.com/1701755/160536955-06b23327-1498-4840-aa23-e499e38dccb6.gif)

</details>

I fixed the `@ symbol` and then it's only the foot note issue (ref + actual note not correct) left.

@dschu-lab I think the foot notes are currently not properly read as you did not expect them. I think it's a good addition but it can come through a separate PR. It looks like everything we need is in the rich text data (I read this https://destroytoday.com/blog/footnotes)

2. Review Result
(Sample Check to double check)
https://satellytescommain21751-featimplementcontentfulsource.gtsb.io/blog/gather-metaverse-bump-each-other-2021/

The video description is missing the link. Is this even supported? 

![Screenshot 2022-03-29 at 07 13 58](https://user-images.githubusercontent.com/1701755/160538212-9f197af1-897b-4389-9ec8-97fd6d4323d1.png)
![Screenshot 2022-03-29 at 07 15 44](https://user-images.githubusercontent.com/1701755/160538241-d4263fe5-cce0-4148-be96-03e0714d0ed0.png)

I also checked a few other random samples and I only found one missing space. So I'm confident we are good.



## Pending Issues
So we can publish and have those tasks left:

+ Images are not limited to their natural size
+ Footnote Handling
+ The video/image description support for links, possible or not ?

@milljoniaer did you notice anything else?
@dschu-lab let's merge or what's left ? 🤩

milljoniaer commented:
No, I dont have noticed anything else. Everything is already covered 👍
